### PR TITLE
BUG Let any user call LoginSessionController

### LINKED ISSUE #45: Refactor http response codes + unit tests to LoginSessionController
Code is currently missing http response codes and specific unit tests

### ACs
- [x] Remove success boolean and replace with http response codes
- [x] Ensure LoginSessionController is covered by unit tests (it currently has none)

### Notes
Missing http response codes as mentioned in [this pr](https://github.com/silverstripe/silverstripe-session-manager/pull/39)
- https://github.com/silverstripe/silverstripe-session-manager/pull/39/files#r602992531
- https://github.com/silverstripe/silverstripe-session-manager/pull/39/files#r602992798
- https://github.com/silverstripe/silverstripe-session-manager/pull/39/files#r602992870
- https://github.com/silverstripe/silverstripe-session-manager/pull/39/files#r602993036

### Pull request
- [ ] https://github.com/silverstripe/silverstripe-session-manager/pull/68

### LINKED ISSUE #67: Unable to revoke non-admin users
Non-admin user isn't able to revoke their own sessions.

With a test user in the group "Content authors", when attempting to revoke a session I see a red toast "Could not log out of session. Try again later"

Originally noted on this pull-request https://github.com/silverstripe/silverstripe-session-manager/pull/62#issuecomment-823742134

### ACs
- [x] Non-admin users are able to revoke their own sessions
- [x] Admin users are able to revoke their own sessions
- [x] Behat and cucumber studio test are updated to require CMS users with restricted access.
- [x] Logout end point is covered by functional test to make sure anyone can terminate sessions and the CSRF token is enforced

# Pull request
- [ ] https://github.com/silverstripe/silverstripe-session-manager/pull/68
- [x] https://studio.cucumber.io/projects/161118/test-plan/folders/1850251/scenarios/6076158 (Updated scenario in cucumber)

### PR DESCRIPTION
**Pull request by maxime-rainville**
Fixes #67 and fixes #45

This is a work in progress. Still need to add the following:
- [x] Test deleting a session you don't have access to
- [x] Test deleting a session that does not exist
- [x] Test an anonymous user
- [x] Update Behat test to use a none-admin account

### CODE REVIEW

File: tests/php/Control/LoginSessionControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        $sessionID = $this->idFromFixture(LoginSession::class, 'x1');
+        $this->session = new Session(['activeLoginSession' => $sessionID]);
+        $this->request =  $this->buildRequestMock('/', [], [], null, $this->session);
+        $this->request->method('getIP')->willReturn('192.168.0.1');
+
+        $memberID = $this->idFromFixture(Member::class, 'owner');
+        $this->logInAs($memberID);
+        SecurityToken::enable();
+        $this->session->set('SecurityID', SecurityToken::inst()->getValue());
```
Comment (maxime-rainville) [inline tests/php/Control/LoginSessionControllerTest.php]: Mostly copied from LoginSessionMiddlewareTest.

File: tests/php/Control/LoginSessionControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        $response = Director::test(
+            "admin/loginsession/remove/$sessionID?SecurityID=$token",
+            null,
+            $this->session,
+            'DELETE'
+        );
+
+        $this->assertEquals('200', $response->getStatusCode());
+        $body = json_decode($response->getBody(), true);
+        $this->assertEquals(['success' => false, 'message' => 'Something went wrong.'], $body);
```
Comment (maxime-rainville) [inline tests/php/Control/LoginSessionControllerTest.php]: At the risk flogging a dead horse, these tests would be easier to write and much more expressive if we were returning sensible HTTP response codes.

For example, we could be catching the wrong error here. Maybe our fake CSRF token was accepted and we are getting an error response because of some other problem.

File: client/src/components/LoginSession/LoginSessionContainer.js [hunk 3]
```diff
@@ -35,23 +35,24 @@ function LoginSessionContainer(props) {
...
+                }
+                props.displayToastFailure(response.message);
+                return Promise.resolve();
+            }).catch(() => {
+                props.displayToastFailure(i18n._t(
+                    'SessionManager.COULD_NOT_LOGOUT',
+                    'Could not log out of session. Try again later.'
+                ));
+            });
+        }).finally(() => {
```
Comment (maxime-rainville) [inline client/src/components/LoginSession/LoginSessionContainer.js]: I refactored this to use proper error handling. Catch statement will handle any scenario where an HTTP Error code is returned.

File: package.json [hunk 4]
```diff
@@ -33,14 +33,16 @@
...
     "babel-jest": "^23.6.0",
     "babel-plugin-transform-runtime": "^6.23.0",
     "babel-runtime": "^6.26.0",
     "enzyme": "^3.3.0",
     "enzyme-adapter-react-16": "^1.10.0",
     "jest-cli": "^23.6.0",
     "mockdate": "^3.0.3",
-    "react-test-renderer": "16.8.3"
+    "react-test-renderer": "16.8.3",
+    "webpack-cli": "^4.7.2"
```
Comment (maxime-rainville) [inline package.json]: yarn build wouldn't do its thing unless I added these bits

File: src/Controllers/LoginSessionController.php [hunk 5]
```diff
@@ -6,17 +6,21 @@
...
+/**
+ * Handles session revokation AJAX request
+ */
 class LoginSessionController extends Controller
 {
-    private static $url_segment = 'loginsession';
-
     private static $url_handlers = [
-        'DELETE remove/$ID' => 'remove',
+        'DELETE $ID' => 'remove',
```
Comment (maxime-rainville) [inline src/Controllers/LoginSessionController.php:20]: This makes more sense from REST point of view.

File: src/Controllers/LoginSessionController.php [hunk 6]
```diff
@@ -29,42 +33,48 @@ class LoginSessionController extends Controller
...
+        if (empty(Security::getCurrentUser())) {
+            return $this->jsonResponse(
+                [
                     'success' => false,
-                    'message' => $failureMessage
-                ]);
-            }
+                    'message' => _t(
+                        __CLASS__ . '.YOUR_SESSION_HAS_EXPIRED',
+                        'Your session has expired'
```
Comment (maxime-rainville) [inline src/Controllers/LoginSessionController.php:42]: I'm adding a check to make sure the user is logged in when they try to invalidated a session.

@silverstripeux "Your session has expired" will be displayed to the user when they try to invalidate a session while their current session has expired. Let me know if you want a different wording for this.

File: src/Controllers/LoginSessionController.php [hunk 7]
```diff
@@ -29,42 +33,48 @@ class LoginSessionController extends Controller
...
+        if (!SecurityToken::inst()->checkRequest($request) || empty($id) || !is_numeric($id)) {
+            return $this->jsonResponse(
+                [
                     'success' => false,
-                    'message' => $failureMessage
-                ]);
-            }
+                    'message' => _t(
+                        __CLASS__ . '.INVALID_REQUEST',
+                        'Invalid request'
```
Comment (maxime-rainville) [inline src/Controllers/LoginSessionController.php:56]: I'm adding an explicit check to make sure ID makes sense. In practice there's no way a user should be able to reach this  during normal operation.

File: src/Controllers/LoginSessionController.php [hunk 8]
```diff
@@ -29,42 +33,48 @@ class LoginSessionController extends Controller
...
-                ]);
-            }
-        } catch (Exception $e) {
-            return $this->jsonResponse([
-                'success' => false,
-                'message' => $failureMessage
-            ]);
+                    'message' => _t(
+                        __CLASS__ . '.COULD_NOT_LOG_OUT_OF_THIS_SESSION',
+                        'Could not log out of this session. Maybe it has already been terminated.'
```
Comment (maxime-rainville) [inline src/Controllers/LoginSessionController.php]: You will get this message if you try to revoke a session that no longer exist. Typical scenario where you would get this is when you have already logout of the other device after loading the "Authenticated devices" form field.

@silverstripeux Let me know if you want to tweak this wording

File: src/FormFields/SessionManagerField.php [hunk 9]
```diff
@@ -118,7 +119,7 @@ public function getSchemaDataDefaults()
      */
     protected function getLoginSessions(Member $member)
     {
-        $logOutEndpoint = LoginSessionController::singleton()->Link('remove');
+        $logOutEndpoint = Director::baseURL() . LoginSessionController::singleton()->Link();
```
Comment (maxime-rainville) [inline src/FormFields/SessionManagerField.php]: This might not actually be needed. It was doing weird thing ... but that might have been my misconfigured route

File: src/Controllers/LoginSessionController.php (hunk 8 above)
Comment (clarkepaul) [reply src/Controllers/LoginSessionController.php]: Maybe this is enough?
This session could not found or is no longer active.

If the above is not enough then we could clarify but it might be too much text? 
It may have already been terminated or timed out.

**Comment by clarkepaul** [comment]
@maxime-rainville Do we test the use of "keep me signed in for 30 days"?

- Is not selected, session only lasts xyz hours/days
- Is selected, remains active for the full 30 days?

File: _config/routes.yml [hunk 10]
```diff
@@ -4,4 +4,4 @@ After: '#coreroutes'
 ---
 SilverStripe\Control\Director:
   rules:
-    'loginsession//remove/$ID': SilverStripe\SessionManager\Controllers\LoginSessionController
+    'loginsession//$ID': SilverStripe\SessionManager\Controllers\LoginSessionController
```
Comment (maxime-rainville) [inline _config/routes.yml:7]: Having a `remove' in your REST API URL makes no sense.

File: client/src/components/LoginSession/LoginSessionContainer.js [hunk 11]
```diff
@@ -34,27 +34,32 @@ function LoginSessionContainer(props) {
...
     })
-      .then(response => {
-        const failure = !response.success;
-        setFailed(failure);
-
-        if (failure) {
-          props.displayToastFailure(response.message);
-        } else {
-          props.displayToastSuccess(response.message);
+    .then(response => {
```
Comment (maxime-rainville) [inline client/src/components/LoginSession/LoginSessionContainer.js:37]: I've reorganised this so we always use HTTP error code. The main `then` statement will only be called when success status code is attached to our response

File: client/src/components/LoginSession/LoginSessionContainer.js [hunk 12]
```diff
@@ -34,27 +34,32 @@ function LoginSessionContainer(props) {
...
+      props.displayToastSuccess(response.message);
+    })
+    .catch(err => {
```
Comment (maxime-rainville) [inline client/src/components/LoginSession/LoginSessionContainer.js:40]: When we have error response, we check if we can parse the body to a JSON object with a message key

File: client/src/components/LoginSession/LoginSessionContainer.js [hunk 13]
```diff
@@ -34,27 +34,32 @@ function LoginSessionContainer(props) {
...
-        ));
-      })
-      .finally(() => {
-        setComplete(true);
-        setSubmitting(false);
+        props.displayToastFailure(response.message);
+        return Promise.resolve();
       });
+    })
+    .catch(() => {
```
Comment (maxime-rainville) [inline client/src/components/LoginSession/LoginSessionContainer.js:52]: If all else fails, we display a generic error message

File: client/src/components/LoginSession/tests/LoginSessionContainer-test.js [hunk 14]
```diff
@@ -118,10 +114,9 @@ describe('LoginSessionContainer', () => {
       submitting: true
     });
 
-    httpResolve({
-      error: true,
-      success: false,
-      message: 'horrible failure'
+    httpReject({
+      response: {
```
Comment (maxime-rainville) [inline client/src/components/LoginSession/tests/LoginSessionContainer-test.js:118]: This mock an Error response. In a perfect world `response` would be returning a `Blob`, but Node 10 doesn't have that API (browsers do).

Manually creating a `text` method that returns a Promise is close enough. I've tested this in IE10 to make sure it works.

File: src/Controllers/LoginSessionController.php [hunk 15]
```diff
@@ -88,6 +94,7 @@ private function jsonResponse(array $response, int $code = 200): HTTPResponse
     {
         return HTTPResponse::create(json_encode($response))
             ->addHeader('Content-Type', 'application/json')
+            ->addHeader('Cache-Control', 'no-cache')
```
Comment (maxime-rainville) [inline src/Controllers/LoginSessionController.php]: Added this extra bit to be 100% we never cache the response.

File: tests/php/Control/LoginSessionControllerTest.php [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+            'DELETE'
+        );
+
+        $this->assertEquals(
+            '200',
+            $response->getStatusCode(),
+            'User can invalidate their session with a valid CSRF token'
+        );
+        $body = json_decode($response->getBody(), true);
+        $this->assertEquals(['message' => 'Successfully logged out of device.'], $body);
```
Comment (maxime-rainville) [inline tests/php/Control/LoginSessionControllerTest.php]: I'm a bit ambivalent about asserting the exact message being returned for all these scenarios. I'm thinking maybe asserting the response code is good enough.

This will break the test if the user has set up their site to a non-English language ... while not terrible, it's not great either.

File: src/Controllers/LoginSessionController.php (hunk 8 above)
Comment (maxime-rainville) [reply src/Controllers/LoginSessionController.php]: Updated the wording 
image

File: src/Controllers/LoginSessionController.php (hunk 6 above)
Comment (maxime-rainville) [reply src/Controllers/LoginSessionController.php:42]: Here's what it looks like. 
image

**Comment by maxime-rainville** [comment]
I've rebased the PR ... just need to tweak the behat test.

> Do we test the use of "keep me signed in for 30 days"?
>     * Is not selected, session only lasts xyz hours/days
>     * Is selected, remains active for the full 30 days?

@clarkepaul This PR is specifically about handling what happens when you use the session manager file to invalidate a session. If you try to logout of a "Remember me session" that has experienced, two things might happen:
- if the login session has been "garbage collected", you will get the "This session could not be found or is no longer active" message
  - In most cases, you will never see those expired sessions unless you somehow manage to load the page in-between the session expiring and the garbage collector doing its job
- if login session garbage collection has not been set up and you try to invalidate an expired session, it will delete the matching LoginSession object and you'll get a success message.

Is that what you wanted to know?

File: src/Controllers/LoginSessionController.php (hunk 15 above)
Comment (emteknetnz) [reply src/Controllers/LoginSessionController.php]: Would `HTTPCacheControlMiddleware::singleton()->disableCache()` work in this context?  If so that's preferable since it's the standard mechanism for this

Other wise the value here should be `no-store` instead which means "The response may not be stored in any cache" - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

File: tests/php/Control/LoginSessionControllerTest.php [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+    /** @var HttpRequest */
+    private $request;
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        $sessionID = $this->idFromFixture(LoginSession::class, 'x1');
+        $this->session = new Session(['activeLoginSession' => $sessionID]);
+        $this->request =  $this->buildRequestMock('/', [], [], null, $this->session);
```
Comment (emteknetnz) [inline tests/php/Control/LoginSessionControllerTest.php]:
```suggestion
        $this->request = $this->buildRequestMock('/', [], [], null, $this->session);
```

File: client/src/components/LoginSession/tests/LoginSessionContainer-test.js [hunk 18]
```diff
@@ -118,10 +114,9 @@ describe('LoginSessionContainer', () => {
...
+        text: () => Promise.resolve(JSON.stringify({ success: false, message: 'horrible failure' })) }
```
Comment (emteknetnz) [inline client/src/components/LoginSession/tests/LoginSessionContainer-test.js]: we no longer need `success: false` since that's been superseded by status codes?

File: src/Controllers/LoginSessionController.php [hunk 19]
```diff
@@ -29,50 +33,52 @@ class LoginSessionController extends Controller
...
-        $failureMessage = _t(__CLASS__ . '.REMOVE_FAILURE', 'Something went wrong.');
-        try {
-            // Ensure CSRF protection
-            if (!SecurityToken::inst()->checkRequest($request)) {
-                return $this->jsonResponse([
-                    'success' => false,
-                    'message' => $failureMessage
-                ]);
-            }
+        if (empty(Security::getCurrentUser())) {
```
Comment (emteknetnz) [inline src/Controllers/LoginSessionController.php:37]: Isn't there some sort of protection on /admin that happens before this point?  Or is that simply to customise the message in this specific scenario only?

Seems like this "your session has expired" functionality should live somehwhere more global?

File: src/Controllers/LoginSessionController.php [hunk 20]
```diff
@@ -29,50 +33,52 @@ class LoginSessionController extends Controller
...
-            $loginSession = LoginSession::get()->byID($id);
-            if (!$loginSession) {
-                return $this->jsonResponse([
-                    'success' => false,
-                    'message' => $failureMessage
-                ]);
-            }
+        // Validate CSRF token and ID parameter
+        $id = $request->param('ID');
+        if (!SecurityToken::inst()->checkRequest($request) || empty($id) || !is_numeric($id)) {
```
Comment (emteknetnz) [inline src/Controllers/LoginSessionController.php]: don't need the `empty($id)` check as both `!is_numeric('')` and `!is_numeric(null)` will both return true

File: src/Controllers/LoginSessionController.php [hunk 21]
```diff
@@ -29,50 +33,52 @@ class LoginSessionController extends Controller
...
+            // We're not making a difference between a non-existent session and session you can not revoke
+            // to prevent an adversary from scanning LoginSession IDs
+            return $this->jsonResponse(
+                [
+                    'message' => _t(
+                        __CLASS__ . '.SESSION_COULD_NOT_BE_FOUND_OR_NO_LONGER_ACTIVE',
+                        'This session could not be found or is no longer active.'
+                    )
+                ],
+                404
```
Comment (emteknetnz) [inline src/Controllers/LoginSessionController.php:74]: It's not really a 404 if it fails the `!$loginSession->canDelete()` check?  404 probably isn't correct on a DELETE endpoint anyway.  Is there something else this could be changed to?

File: tests/behat/features/revoke.feature [hunk 22]
```diff
@@ -6,16 +6,13 @@ Feature: See other devices and revoke their access
   So that I can revoke their access
 
   Background:
-    Given I am logged in with "ADMIN" permissions
```
Comment (emteknetnz) [inline tests/behat/features/revoke.feature:9]: Worth having a separate test to ensure that ADMIN cannot revoke AUTHOR

OK to create a new issue for this instead of fixing on this PR

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: client/src/components/LoginSession/tests/LoginSessionContainer-test.js (hunk 18 above)
Comment (maxime-rainville) [reply client/src/components/LoginSession/tests/LoginSessionContainer-test.js]: Yep

File: src/Controllers/LoginSessionController.php (hunk 19 above)
Comment (maxime-rainville) [reply src/Controllers/LoginSessionController.php:37]: It's a regular controller now, so it's no longer showing under `admin`. I guess we could put a restriction on the action, but we wouldn't got a nice error message in that scenario.

File: src/Controllers/LoginSessionController.php (hunk 21 above)
Comment (maxime-rainville) [reply src/Controllers/LoginSessionController.php:74]: If we return a 403, an attacker could use this to try to scan for valid session LoginSession ID. There's a comment to this affect.

File: tests/behat/features/revoke.feature (hunk 22 above)
Comment (maxime-rainville) [reply tests/behat/features/revoke.feature:9]: Not sure about this one ... I think behat tests should mostly be aimed at testing the "happy path" and avoid testing edge cases.

File: src/Controllers/LoginSessionController.php (hunk 21 above)
Comment (maxime-rainville) [reply src/Controllers/LoginSessionController.php:74]: My understanding is that when trying to do something to a resource that doesn't exist, 404 is the generally accepted response.

A quick look through stack overflow suggest that some people will also return a '204 No Content' (which is bodyless and wouldn't let us return a message). Another option is a '410 Gone' if you're trying to delete something that once existed but has already been deleted ... which we don't have a way of determining here.

So in summary, I think a 404 is the best option here.

File: tests/behat/files/testfile.txt [hunk 23]
```diff
@@ -1 +0,0 @@
-abc
```
Comment (maxime-rainville) [inline tests/behat/files/testfile.txt:1]: This file did not serve any discernible purpose so I nuked it.

File: src/Controllers/LoginSessionController.php (hunk 15 above)
Comment (maxime-rainville) [reply src/Controllers/LoginSessionController.php]: Very good suggestion ... done. I also added a unit test to validate tho "Cache-Control" header.

File: src/Controllers/LoginSessionController.php (hunk 20 above)
Comment (maxime-rainville) [reply src/Controllers/LoginSessionController.php]: Fixed.

File: client/src/components/LoginSession/tests/LoginSessionContainer-test.js (hunk 18 above)
Comment (emteknetnz) [inline client/src/components/LoginSession/tests/LoginSessionContainer-test.js]:
```suggestion
        text: () => Promise.resolve(JSON.stringify({ message: 'horrible failure' })) }
```
We should remove it then? :-)

File: src/Controllers/LoginSessionController.php [hunk 24]
```diff
@@ -6,17 +6,22 @@
 use SilverStripe\Control\Controller;
 use SilverStripe\Control\HTTPRequest;
 use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\Middleware\HTTPCacheControlMiddleware;
+use SilverStripe\Security\Security;
 use SilverStripe\Security\SecurityToken;
 use SilverStripe\SessionManager\Models\LoginSession;
 
+/**
+ * Handles session revokation AJAX request
```
Comment (emteknetnz) [inline src/Controllers/LoginSessionController.php]:
```suggestion
 * Handles session revocation AJAX request
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Looks great. A couple of trivial changes

**Review by emteknetnz — APPROVED** [review]
MOG

**Comment by michalkleiner** [comment]
What does MOG stand for @emteknetnz? Might be confusing for non-native speakers following/using Silverstripe when acronyms like that are used.

**Comment by emteknetnz** [comment]
Merge on green - it's a workflow thing

We use it when we want to approve code changes though travis has not yet finished running.  If the build goes green then anyone with merge permissions can merge the PR, including the person who originally created the PR

**Comment by maxime-rainville** [comment]
While in theory, we would like to have the community be involved in every PR and issue, the unfortunate reality is that there's a number of us that work on the project full time and that we will inevitably develop our own lingo.

I don't think it's practical for us to manage every PR in way that any community member could readily jump on it and understand every little detail of the conversation. But it's a good reminder that a lot of our conversations are not always accessible or self-evident, and that we should look for opportunities to make things easier to understand.
